nomad/search_endpoint.go

```diff
@@ -229,3 +491,140 @@ func (s *Search) PrefixSearch(args *structs.SearchRequest, reply *structs.Search
 		}}
 	return s.srv.blockingRPC(&opts)
 }
+
+// FuzzySearch is used to list fuzzy or prefix matches for a given text argument and Context.
+// If the Context is "all", all searchable contexts are searched. If ACLs are enabled,
+// results are limited to policies of the provided ACL token.
+//
+// These types are limited to prefix UUID searching:
+//   Evals, Deployments, ScalingPolicies, Volumes
+//
+// These types are available for fuzzy searching:
+//   Nodes, Namespaces, Jobs, Allocs, Plugins
+//
+// Jobs are a special case that expand into multiple types, and whose return
+// values include Scope which is a descending list of IDs of parent objects,
+// starting with the Namespace. The subtypes of jobs are fuzzy searchable.
+//
+// The Jobs type expands into these sub types:
+//   Jobs, Groups, Services, Tasks, Images, Commands, Classes
+//
+// The results are in descending order starting with strongest match, per Context type.
+func (s *Search) FuzzySearch(args *structs.FuzzySearchRequest, reply *structs.FuzzySearchResponse) error {
+	if done, err := s.srv.forward("Search.FuzzySearch", args, args, reply); done {
+		return err
+	}
+	defer metrics.MeasureSince([]string{"nomad", "search", "fuzzy_search"}, time.Now())
+
+	// check that fuzzy search API is enabled
+	if !s.srv.config.SearchConfig.FuzzyEnabled {
+		return fmt.Errorf("fuzzy search is not enabled")
+	}
+
+	// check the query term meets minimum length
```

Comment:
The ACL check should be before argument validation so that we return ACL errors first. Probably before config validation too?

Reply:
Yup, I agree. One thing though - to make `namespace=*` work, this up front ACL check is basically bypassed, instead defering to iterators wrapped with filters, doing the same ACL checks removing objects not allowed by the given token. So this is really more of a "at least try to give them an error if they specify a namespace they can't access" optimization.